### ISSUE TITLE
Date picker field improvements

### ISSUE THREAD

**Issue by PVince81** [owncloud/phoenix#2258]
- [ ] add placeholder when empty
- [ ] add date picker icon inside the field

- [ ] how about accessibility ? @marcus-herrmann 

<img width="35" alt="image" src="https://user-images.githubusercontent.com/277525/66812333-ffc4c680-ef32-11e9-9570-e6fa2be4a3c5.png">

**Comment by PVince81** [comment]
<img width="674" alt="image" src="https://user-images.githubusercontent.com/277525/66812368-0e12e280-ef33-11e9-90d7-02489d767766.png">

**Comment by marcus-herrmann** [comment]
Date Pickers and accessibility, classic topic.

# Current implementation
The current implementation in inaccessible:
* It is not usable with keyboard only
* It has no focus management, focus is not sent into the container that appears to be a modal, nor sending it back to the trigger on close
* the date picker "modal" is not a real one (see https://github.com/owncloud/phoenix/issues/2263#issuecomment-543516672)
* the date picker is unusable with screen readers and voice control software (like Dragon Speech)
* opening the "modal" on the focus event is an unexpected change of context, therefore could be considered as a violation of [WCAG 3.2.2](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#consistent-behavior-unpredictable-change))

Web accessibility expert [Adrian Roselli has a great article on the whole topic](https://adrianroselli.com/2019/07/maybe-you-dont-need-a-date-picker.html) (bordering UX, maybe relevant for @wuenschedesign), unfortuately there is something wrong with the embedded Codepen.

# In-depth article
Takeaways from the article:
* Datepickers can be used for indicating a precise date or a time span. Different use cases, different accessibility concerns. **Can we be sure that the date picker in Owncloud will be only for precise dates and not spans of time?**
* JS Datepickers, even them who claim to be accessible, are not
* Adrian did 20 years of user research on this topic and noted high amounts of frustration with users of assistive technologies (e.g. Screen readers, Voice Control software)
* People don't want to learn an interface, but enter the data, "get stuff done"
* Even W3C's Web Accessibility Initiative does not dare to recommend a best practice
* Even the browser's native datepickers, `<input type=date>`, are flawed and inaccessible (!)
* Validate client-side, quickly and obviously, and of course, on the server side as well.

# So what to do
A way out of this dilemma could be mentioned [in the article's comments](https://adrianroselli.com/2019/07/maybe-you-dont-need-a-date-picker.html#comment-159384). My interpretation of it:

* Supplying a text field (`type="text"`!)
* Putting a button adjacent to the field that triggers a date picker
* Adding instructions to the text field (via `aria-describedby`– I could demonstrate if you are interested) regarding what date input format is requested
* Making sure that the date picker is at least keyboard accessible.

If you are interested I could look into how big e-commerce/travel websites solve this (at least the ones that a known for good accessibility, like Expedia)

**Comment by PVince81** [comment]
> If you are interested I could look into how big e-commerce/travel websites solve this (at least the ones that a known for good accessibility, like Expedia)

Sounds like a plan.

So it sounds like we might need to swap the date picker components.

How about detecting screen reader scenarios and swapping the date picker with a simple text field ? Or can we make the date picker field invisible to screen reader and have another "visually hidden" field that is considered present for screen readers ?

Is detecting screen readers and switching rendering mode an anti-pattern ?

**Comment by marcus-herrmann** [comment]
> Is detecting screen readers and switching rendering mode an anti-pattern?

If you take this literaly – it is, at least [ethically](https://tink.uk/thoughts-on-screen-reader-detection/).

But hiding a (sr-)broken experience from screen readers **if** we could supply an equivalent experience could work, but I have to do more research on that. 

One strategy could be (single date, no timespan selection here):

* Hide the datepicker-triggering button with `aria-hidden="true"`(which is, in some regards, the opposite of visually hidden)
* But then describe what form and format of input you expect (like described above, with`aria-describedby`).

But to re-iterate: This would only deal with screen readers. We have have to take other things in consideration:

* How the date picker overlay would work with Speech recognition software (Dragon, Apple's Voice Control)
* That it's keyboard accessible for keyboard-only or switch input users
* That it can be enlargened without visually by zooms of all sorts without breaking (this concern actually is valid for the whole Phoenix interface)

I don't properly suggest this strategy yet because it is a) an assumption and b) I haven't yet researched the side effects on other forms of assitive technology/accessibility. I would extend my research with that.

**Comment by PVince81** [comment]
So far I'd say the date picker is only here as a convenience. In general all we want is that the user inputs some date, but typing a date can be cumbersome due to the different formats.

This is where I'm wondering whether such software should have a way to provide their own method for inputting dates. Not sure if it's their responsibility.
This then brought me to the "input type='date'" element https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date which seems to be a way for browsers to display a date picker and even supports limiting ranges (the latter is needed by OC in some scenarios).

Is there anything related to "input type='date'" for accessibility tools ?

So far it sounds like more research is needed. Maybe as you suggested, have a look how some popular accessible websites are solving it.

**Comment by marcus-herrmann** [comment]
Like mentioned above, `<input type="date">` is <tldr>frustratingly not generally accessible</tldr> ([long read](https://www.hassellinclusion.com/blog/input-type-date-ready-for-use/)

**Comment by marcus-herrmann** [comment]
# Research
## Accessible solution in the wild
### Expedia

Homepage, https://www.expedia.com/

- When the "Check in" or "Check out" text input is focused, the date picker overlay is open and the accessible name of the text input changes from its label ("Check-In", supplied by the `<label>`) to a rather verbose text, provided by aria label attribute, which is only applied when its focused: "Check-in Check-in expanded. Enter a date in mm/dd/yyyy format or choose one from the table below. Showing Oct 2019 Nov 2019". I'm not quite sure why they did not use `aria-describedby` the same way. Maybe user testing
- If you are wondering why there's repetition at the beginning, "Check-in Check-in expanded. Enter a date in mm/dd/yyyy format or choose one from the table below. Showing Oct 2019 Nov 2019", I'm certain it's because of WCAG 2.5.3, "Label in name" ([explainer]([https://knowbility.org/blog/2018/WCAG21-253LabelInName/](https://knowbility.org/blog/2018/WCAG21-253LabelInName/)))
- Date picker uses the concept of the disclosure widget and it is activated/expanded on focus of "Check in", "Check out" text inputs (since it's no modal and no change of context, an activation on focus is okay here)
- The markup of the date picker consists (mostly) of a data/date table where inaccessible dates (the ones in the past) are visually perceivable, but greyed out, and non-interactive. This is ensured with the `disabled` attribute on the single days, which are buttons. In addition to that, inactive days are completly hidden to screen reader users, with `aria-hidden="true"`). So what a screen reader user perceives is one or two tables consisting of only available, selectable, reachable dates. Furthermore, the weekday indicator (S M T W T F) is perceived with full weekday names, supplied by using visually hidden elements. In addition the table is captioned and marked up properly with `<thead>`, `<tbody>`, etc.

**Comment by marcus-herrmann** [comment]
My recommendation would be right now to try to implement a solution inspired by booking.com's date picker. Provinding select input options prevent unclear date inputs (that could happen with a freetext input, even with a good placeholder example). Also, it would make validation easier since the system could offer only selectable dates. This would make a (select based) input a valid input (at least on the client side). But we have to make sure that the whole date picker widget for mouse, touch, voice input users is hidden for screen readers (`aria-hidden="true"`).

Same goes for the "visual" date picker. Only offering selectable dates (present or future) will lead to valid inputs (still - at least on the client side. But I guess a server side validation of these dates in already in place).

Bear in mind that this recommendation is based on the research above but needs to be tested with real users and real assistive technology.

**Comment by wuenschedesign** [comment]
To solve the problem with accessibility
![Add-Public-Link-05](https://user-images.githubusercontent.com/40062503/67288560-7fc3d100-f4dd-11e9-8951-e3f5f3ca585f.png)
